src/main/java/io/castle/example/LoginServlet.java

```diff
@@ -1,38 +1,83 @@
 
 @WebServlet("/login")
 public class LoginServlet extends HttpServlet {
 
+    private static final Logger LOGGER = Logger.getLogger(LoginServlet.class.getName());
+
     @Override
     protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
-        Castle.setAPISecret(Config.apiKey());
+        CastleApi castleApi = Castle.sdk().onRequest(req);
 
-        Event event = new Event();
-        event.setUserId("1234");
-        event.setName("$login.succeeded");
+        String overrideString = req.getParameter("override");
+        AuthenticateAction override = null;
 
-        UserInfoHeader userInfoHeader = UserInfoHeader.fromRequest(req);
+        if (overrideString != null && overrideString.equals("challenge")) {
+            override = AuthenticateAction.CHALLENGE;
+        } else if (overrideString != null && overrideString.equals("deny")) {
+            override = AuthenticateAction.DENY;
+        }
 
-        Event result = Event.setUserInfoHeader(userInfoHeader).authenticate(event);
+        String username = req.getParameter("username");
+        String password = req.getParameter("password");
 
-        if (result.getAction().equals("challenge")) {
-            // challenge
-        } else {
-            // 'allow' or 'deny'
-        }
+        HttpSession session = req.getSession(true);
 
-        resp.sendRedirect("authenticated.jsp");
+        TestUser user = UserAuthenticationBackend.findUser(username);
+        if (user != null && user.getPassword().compareTo(password) == 0) {
+            String id = user.getId().toString();
+            Verdict verdict = castleApi.authenticate("$login.succeeded", id);
+            if (override != null) {
+                verdict.setAction(override);
+                LOGGER.info("The result of the Castle API authenticate call has been overridden to: "
+                        + override.toString());
+            }
+            switch (verdict.getAction()) {
+                case DENY: {
+                    castleApi.track("$login.failed", id, user);
+                    session.invalidate();
+                    resp.sendRedirect("authentication_error.jsp");
+                }
+                break;
+                case ALLOW: {
+                    castleApi.track("$login.succeeded", id);
+                    session.setAttribute("currentSessionUser", user);
+                    resp.sendRedirect("/");
+                }
+                break;
+                case CHALLENGE: {
+                    castleApi.track("$challenge.requested", id, user);
+                    session.setAttribute("challengedUser", user);
+                    resp.sendRedirect("challenge.jsp");
+                }
+                break;
+            }
+        } else
+
+        {
+            if (user != null) {
+                castleApi.track("$login.failed", user.getId().toString());
+            } else {
+                castleApi.track("$login.failed");
```

Comment:
add the properties to login failed. see ruby example
```ruby
castle.track(
  name: '$login.failed',
  user_id: '1234', # Omit this if no matching user was found
  properties: {
    email: 'johan@castle.io'
  })
```

Reply:
in case of `user == null`

Reply:
done.